src/libprojectM/PCM.cpp

```diff
@@ -24,381 +24,528 @@
+        {
+            float g = pow(M_E, i * i * k);
+            freq[i * 2] *= g;
+            freq[i * 2 + 1] *= g;
+        }
+        freq[1] *= pow(M_E, FFT_LENGTH*FFT_LENGTH*k);
+    }
+    else
+    {
+        // butterworth
+        // this might be slightly faster to compute. pow() is expensive
+        double k = 1.0 / ((1 - smoothing) * (1 - smoothing) * FFT_LENGTH * FFT_LENGTH);
+        for (int i = 1; i < FFT_LENGTH; i++)
+        {
+            float b = 1.0 / (1.0 + (i * i * k));
+            freq[i * 2] *= b;
+            freq[i * 2 + 1] *= b;
+        }
+        freq[1] *= 1.0 / (1.0 + (FFT_LENGTH*FFT_LENGTH*k));
+    }
+
+    // inverse fft
+    rdft(FFT_LENGTH*2, -1, freq, ip, w);
+    for (size_t j = 0; j < FFT_LENGTH*2; j++)
+        freq[j] *= 1.0 / FFT_LENGTH;
+
+    // copy out with zero-padding if necessary
+    size_t count = samples<FFT_LENGTH ? samples : FFT_LENGTH;
+    for (size_t i=0 ; i<count ; i++)
+        data[i] = freq[i%(FFT_LENGTH*2)];
+    for (size_t i=count ; i<samples ; i++)
+        data[i] = 0;
+}
 
-void PCM::addPCM16Data(const short* pcm_data, short samples)  {
-   int i, j;
 
-   for (i = 0; i < samples; ++i) {
-      j=i+start;
-      PCMd[0][j % maxsamples]=(pcm_data[i * 2 + 0]/16384.0);
-      PCMd[1][j % maxsamples]=(pcm_data[i * 2 + 1]/16384.0);
-   }
+/* NOTE: Still don't have real support for smoothing parameter, but this gets close to the milkdrop2 default look */
+void PCM::getSpectrum(float *data, int channel, size_t samples, float smoothing)
+{
+    assert(channel == 0 || channel == 1);
+    _updateFFT();
 
-   start = (start + samples) % maxsamples;
+    float *spectrum = channel == 0 ? spectrumL : spectrumR;
+    if (smoothing == 0)
+    {
+        size_t count = samples <= FFT_LENGTH ? samples : FFT_LENGTH;
+        for (size_t i = 0; i < count; i++)
+            data[i] = spectrum[i];
+        for (size_t i = count; i < samples; i++)
+            data[0] = 0;
+    }
+    else
+    {
+        float l2 = 0, l1 =0 , c = 0, r1, r2;
+        r1 = spectrum[0]; r2 = spectrum[0+1];
+        for (size_t i = 0; i < samples; i++)
+        {
+            l2 = l1;
+            l1 = c;
+            c = r1;
+            r1 = r2;
+            r2 = (i + 2) >= samples ? 0 : spectrum[i + 2];
+            data[i] = (l2 + 4 * l1 + 6 * c + 4 * r1 + r2) / 16.0;
+        }
+    }
+}
 
-   newsamples+=samples;
-   if (newsamples>maxsamples) newsamples=maxsamples;
-     numsamples = getPCMnew(pcmdataR,1,0,waveSmoothing,0,0);
-    getPCMnew(pcmdataL,0,0,waveSmoothing,0,1);
-    getPCM(vdataL,FFT_LENGTH,0,1,0,0);
-    getPCM(vdataR,FFT_LENGTH,1,1,0,0);
+void PCM::_updateFFT()
+{
+    if (newsamples > 0)
+    {
+        _updateFFT(0);
+        _updateFFT(1);
+        newsamples = 0;
+    }
 }
 
+void PCM::_updateFFT(size_t channel)
+{
+    assert(channel == 0 || channel == 1);
+
+    double *freq = channel==0 ? freqL : freqR;
+    _copyPCM(freq, channel, FFT_LENGTH*2);
+    rdft(FFT_LENGTH*2, 1, freq, ip, w);
```

Comment:
So we do rdft on `getPCM` and here both?

Reply:
The call to rdft in getPCM() is actually an inverse fft.  The call here is doing the time->freq transformation.  In getPCM() we filter the high-freq values then do the freq->time conversion.  And voila, smooth!